### PR TITLE
3441 Block file submission wrong date

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes

Pull request closes #3441_

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
1. Choose a year and quarter and submit a file that year and quarter doesn't match
1. See that an error shows year and quarter doesn't match

## Deliverables

 - [x] The system validates the Year/Quarter in the file header before allowing submission.
 - [x] If an incorrect period is detected:
       - [ ] Submission is blocked.
       - [ ] In-line error message appears describing the issue.
       - [ ] Error message includes the correct period detected for submission.
       - [ ] Error message includes a link to the Knowledge Center for further assistance.
 - [x] The user experience is consistent with existing error handling for file extension errors.
 - [x] Users will not be able to press submit until the error goes away
 - [x] State reset/rescan when new file is attached
 - [x] Testing Checklist has been run and all tests pass
 - [ ] README is updated, if necessary

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
 - [x] The system validates the Year/Quarter in the file header before allowing submission.
 - [x] If an incorrect period is detected:
       - [⚠️  ] Submission is blocked.

_yes, except in the case of some types of encoded file types_

       - [x] In-line error message appears describing the issue.
       - [x] Error message includes the correct period detected for submission.
       - [x] Error message includes a link to the Knowledge Center for further assistance.
 - [x] The user experience is consistent with existing error handling for file extension errors.
 - [x] Users will not be able to press submit until the error goes away
 - [x] State reset/rescan when new file is attached
 - [x] Testing Checklist has been run and all tests pass
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [x] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-frontend/src/components/FileUpload/FileUpload.jsx [hunk 1]
```diff
@@ -174,7 +177,30 @@ function FileUpload({ section, setLocalAlertState }) {
...
+        dispatch({
+          type: SET_FILE_ERROR,
+          payload: {
+            error: {
+              message:
+                `File header does not match selected year: ` +
+                selectedYear +
+                ` and quarter: ` +
+                selectedQuarter +
+                ` .`,
```
Comment (jtimpe) [inline tdrs-frontend/src/components/FileUpload/FileUpload.jsx]:
```suggestion
                `.`,
```

File: tdrs-frontend/src/components/FileUpload/utils.jsx [hunk 2]
```diff
@@ -140,3 +140,28 @@ export const tryGetUTF8EncodedFile = async function (fileBytes, file) {
...
+  // reads the first line of file formatted as UTF8Encoded
+  const decoder = new TextDecoder('utf-8') // Or another encoding if needed
+  const textResult = decoder.decode(result)
+  const lines = textResult.split('\n')
+  const firstLine = lines[0]
+  const yearQuarterRegex = '[0-9]{4}[1-4]'
+  const yearQuarter = firstLine.match(yearQuarterRegex)
+  const year = yearQuarter?.[0]?.slice(0, 4)
+  const quarter = 'Q' + yearQuarter?.[0]?.slice(4, 5)
+  if (yearQuarter && (year !== selectedYear || quarter !== selectedQuarter)) {
```
Comment (jtimpe) [inline tdrs-frontend/src/components/FileUpload/utils.jsx]: i don't think this handles the q1/q4 case where the year needs to be incremented (or decremented?). in the backend, we have the `fiscal_to_calendar` and `calendar_to_fiscal` utils. could we replicate the needed logic here?

https://github.com/raft-tech/TANF-app/blob/20d3be07dda6e9f2a384569f9969dd0ad413ec60/tdrs-backend/tdpservice/parsers/util.py#L154

**Review by jtimpe — CHANGES_REQUESTED** [review]
Ran into unexpected errors while testing the file `ADS.E2J.FTP1.TS06` for `2021` `Q1` (which i believe is the correct submission year/quarter for that file) because the header reads `20204`

File: tdrs-frontend/src/components/FileUpload/utils.jsx [hunk 3]
```diff
@@ -140,3 +140,28 @@ export const tryGetUTF8EncodedFile = async function (fileBytes, file) {
...
+  file,
+  selectedYear,
+  selectedQuarter
+) {
+  // reads the first line of file formatted as UTF8Encoded
+  const decoder = new TextDecoder('utf-8') // Or another encoding if needed
+  const textResult = decoder.decode(result)
+  const lines = textResult.split('\n')
+  const firstLine = lines[0]
+  const yearQuarterRegex = '[0-9]{4}[1-4]'
```
Comment (elipe17) [inline tdrs-frontend/src/components/FileUpload/utils.jsx:167]: Do we need regex here? Can we slice the bytes/string?

File: tdrs-frontend/src/components/FileUpload/utils.jsx (hunk 2 above)
Comment (elipe17) [reply tdrs-frontend/src/components/FileUpload/utils.jsx]: Confirmed Jan's intuition. Need some custom logic for the border quarters. The case below shouldn't throw the error.
Screenshot 2025-03-11 at 1 07 15 PM

File: tdrs-frontend/src/components/FileUpload/utils.jsx (hunk 3 above)
Comment (raftmsohani) [reply tdrs-frontend/src/components/FileUpload/utils.jsx:167]: slice works but this is cleaner to me

File: tdrs-frontend/src/components/FileUpload/utils.jsx (hunk 2 above)
Comment (raftmsohani) [reply tdrs-frontend/src/components/FileUpload/utils.jsx]: fixed!

**Comment by raftmsohani** [comment]
working on fixing the failing test..

File: tdrs-frontend/src/components/FileUpload/utils.jsx (hunk 2 above)
Comment (jtimpe) [reply tdrs-frontend/src/components/FileUpload/utils.jsx]: @raftmsohani still not working for me for that file

Comment (raftmsohani) [reply tdrs-frontend/src/components/FileUpload/utils.jsx]: good catch @jtimpe , I see there is a problem with year. I will fix this

**Comment by victoriaatraft** [comment]
Screenshot 2025-03-14 at 12 52 32 PM

**Comment by lhuxraft** [comment]
@raftmsohani Tested in a11y and functionality looks great! However, the language in the mockup does not appear to match what was implemented. Please reference [this mockup](https://www.figma.com/design/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=10614-3038&t=zcFVjWCFkvX9rpf8-0) for the language. The "Need Help?" link should point to [this URL](https://tdp-project-updates.app.cloud.gov/knowledge-center/uploading-data.html#reporting-period) as well (as noted in the mockup).

A couple other notes for consideration:
- When I get the fiscal quarter/year error and then try to click 'Submit', I get another error that specifies that the reason I can't submit is because "no changes were made to data files". @ADPennington @klinkoberstar What do we think about updating this language for this error type? See screenshot:
image
- Another minor bug we want to look into - if I submit a file that is correct and then switch to another file that is incorrect, I lose the ability to pick another file. See screenshot:
image

cc @victoriaatraft @reitermb

**Review by lhuxraft — CHANGES_REQUESTED** [review]
@raftmsohani Tested in a11y and functionality looks great! However, the language in the mockup does not appear to match what was implemented. Please reference [this mockup](https://www.figma.com/design/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=10614-3038&t=zcFVjWCFkvX9rpf8-0) for the language. The "Need Help?" link should point to [this URL](https://tdp-project-updates.app.cloud.gov/knowledge-center/uploading-data.html#reporting-period) as well.

A couple other notes for consideration:
- When I get the fiscal quarter/year error and then try to click 'Submit', I get another error that specifies that the reason I can't submit is because "no changes were made to data files". @ADPennington @klinkoberstar What do we think about updating this language for this error type? See screenshot:
image
- Another minor bug we want to look into - if I submit a file that is correct and then switch to another file that is incorrect, I lose the ability to pick another file. See screenshot:
image

cc @reitermb @victoriaatraft

**Comment by raftmsohani** [comment]
> @raftmsohani Tested in a11y and functionality looks great! However, the language in the mockup does not appear to match what was implemented. Please reference [this mockup](https://www.figma.com/design/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=10614-3038&t=zcFVjWCFkvX9rpf8-0) for the language. The "Need Help?" link should point to [this URL](https://tdp-project-updates.app.cloud.gov/knowledge-center/uploading-data.html#reporting-period) as well.
> 
> A couple other notes for consideration:
> 
> * When I get the fiscal quarter/year error and then try to click 'Submit', I get another error that specifies that the reason I can't submit is because "no changes were made to data files". @ADPennington @klinkoberstar What do we think about updating this language for this error type? See screenshot:
>   image
> * Another minor bug we want to look into - if I submit a file that is correct and then switch to another file that is incorrect, I lose the ability to pick another file. See screenshot:
>   image
> 
> cc @reitermb @victoriaatraft

@lhuxraft fixed the text to read like the figma and added the URL, however, I will need to investigate more about disappearing file upload box

**Review by lhuxraft — APPROVED** [review]
@raftmsohani Thank you for looking into this - deployed to a11y to double check the language and looks great! 💯 

If you're able to determine that the file upload box behavior wasn't introduced by the changes in this PR, I'm fine with just creating a follow-on ticket to address the bug. I can also ask in standup on Monday if anyone is aware of this behavior (possibly related to file input collapse ticket we worked on a few weeks ago?)

**Review by jtimpe — APPROVED** [review]

File: tdrs-frontend/src/components/FileUpload/FileUpload.jsx [hunk 4]
```diff
@@ -174,7 +177,55 @@ function FileUpload({ section, setLocalAlertState }) {
...
+          case '3':
+            error_period = 'Apr 1 - Jun 30, '
+            break
+          case '4':
+            error_period = 'Jul 1 - Sep 30, '
+            break
+          default:
+            error_period = ''
+        }
+        dispatch({
```
Comment (elipe17) [inline tdrs-frontend/src/components/FileUpload/FileUpload.jsx:216]: Noticing a visual bug when the error is produced. I think it might be resolved if we add `createFileInputErrorState` before the dispatch

Screenshot 2025-03-17 at 11 35 36 AM

Comment (elipe17) [reply tdrs-frontend/src/components/FileUpload/FileUpload.jsx:216]: Note the visual bug does seem to be intermittent

**Review by elipe17 — APPROVED** [review]
Left one comment/suggestion. otherwise looks good, LGTM!

**Comment by raftmsohani** [comment]
> Left one comment/suggestion. otherwise looks good, LGTM!

Solved!!!!

**Comment by raftmsohani** [comment]
The test is failing now, I am working on it

**Comment by raftmsohani** [comment]
> The test is failing now, I am working on it

DONE

**Review by victoriaatraft — APPROVED** [review]

**Comment by ADPennington** [comment]
@raftmsohani can you resolve the merge conflict on this branch when you get a chance?

**Review by klinkoberstar — APPROVED** [review]
I tested this with a variety of files for TANF, SSP, and Tribal TANF and this seems to be working as expected: 

Looking at each of the ACs:
- The system validates the Year/Quarter in the file header before allowing submission. ✅ 
- If an incorrect period is detected:
  - Submission is blocked. ✅ 
  - In-line error message appears describing the issue. ✅ 
  - Error message includes the correct period detected for submission.✅ 
  - Error message includes a link to the Knowledge Center for further assistance. ✅ 
- The user experience is consistent with existing error handling for file extension errors. ✅ 
- Users will not be able to press submit until the error goes away ✅ - _Technically users can press submit, but system will return that no new file has been selected. This behavior seems good to me and aligns with the file extension error behavior_
- State reset/rescan when new file is attached ✅ 

If the header date is correct but record dates are not, it will allow you to submit and then the file is rejected. I think this is expected and even preferred b/c we do don't want to process the entire file at the point of submission. 

@ADPennington and I briefly discussed if we should add in the year for the submitted file. We are leaning towards leaving it as is (no year) to reduce confusion - but Alex may address this further in her comments!

This will be a big help for STTs (and OFA)!

File: tdrs-frontend/src/components/FileUpload/utils.jsx [hunk 5]
```diff
@@ -140,3 +140,45 @@ export const tryGetUTF8EncodedFile = async function (fileBytes, file) {
...
+    let quarter = parseInt(calendarQuarter)
+    let year = parseInt(calendarYear)
+    if (quarter === 4) {
+      year = year + 1
+    }
+    quarter = quarter === 4 ? 1 : quarter + 1
+    return [year.toString(), quarter.toString()]
+  }
+  // reads the first line of file formatted as UTF8Encoded
+  const decoder = new TextDecoder('utf-8') // Or another encoding if needed
```
Comment (ADPennington) [inline tdrs-frontend/src/components/FileUpload/utils.jsx:163]: @raftmsohani I'm noticing an edge case where the wrong quarter is allowed to be submitted (but is rejected). its related to the encoding. test file below, which is UTF16LE encoded. 

[2024.Q1.Active Case Data_UTF16LE.txt](https://github.com/user-attachments/files/19455724/2024.Q1.Active.Case.Data_UTF16LE.txt)

**Review by ADPennington — COMMENTED** [review]
@raftmsohani @lhuxraft @ttran-hub @klinkoberstar my test notes are below. this is in good shape!  Let me know:
- if the one issue i flagged [here](https://github.com/raft-tech/TANF-app/pull/3501#discussion_r2012846191) needs to be addressed here or in follow-on ticket. 
- It'd be great to release with this sprint if release notes will be ready this week,  but flexible!

---

test scenarios:

- What happens if the file's first line doesn't contain a valid year/quarter?



> if out of range quarter, then file can be submitted but is rejected because rejected because: `HEADER Item 5 (quarter): 5 is not in [1, 2, 3, 4].` this is okay, because this is a scenario where the user needs to make a correction to the file :heavy_check_mark:

> out of range year, file cannot be submitted because: `File contains data from Jan 1 - Mar 31, which belongs to Fiscal Year 5202, Quarter 2. Adjust your search parameters or upload a different file. Need help?` :heavy_check_mark:

> out of range year + out of range qtr, file can be submitted but rejected because: `HEADER Item 5 (quarter): 9 is not in [1, 2, 3, 4].` this is okay, because this is a scenario where the user needs to make a correction to the file :heavy_check_mark:


- Does the function handle cases where the file content is empty or malformed?

> if empty file, can't submit file because `Request failed with status code 400: The submitted file is empty.` ✔️

> if no header, then file can be submitted but is rejected because `Your file does not begin with a HEADER record.` :heavy_check_mark:

- How does it handle multiple headers (with different qtrs)
> file will be handled according to the first record in the file ✔️ 

- How does it handle different file encodings? [reference](https://github.com/raft-tech/TANF-app/pull/3438#pullrequestreview-2601855052))

> see [here](https://github.com/raft-tech/TANF-app/pull/3501#discussion_r2012846191) for edge case that previously worked ⚠️

File: tdrs-frontend/src/components/FileUpload/utils.jsx (hunk 5 above)
Comment (raftmsohani) [reply tdrs-frontend/src/components/FileUpload/utils.jsx:163]: Since this is covering almost 95% of the cases, I suggest we go ahead and then create a follow up ticket to add other encodings. We should be able to utilize existing utilities function

Comment (ADPennington) [reply tdrs-frontend/src/components/FileUpload/utils.jsx:163]: sure. could you spin up the ticket so I can approve this with links to followup @raftmsohani ?

**Comment by lhuxraft** [comment]
> @raftmsohani @lhuxraft @ttran-hub @klinkoberstar my test notes are below. this is in good shape! Let me know:
> 
> * if the one issue i flagged [here](https://github.com/raft-tech/TANF-app/pull/3501#discussion_r2012846191) needs to be addressed here or in follow-on ticket.
> * It'd be great to release with this sprint if release notes will be ready this week,  but flexible!
> 
> test scenarios:
> 
> * What happens if the file's first line doesn't contain a valid year/quarter?
> 
> > if out of range quarter, then file can be submitted but is rejected because rejected because: `HEADER Item 5 (quarter): 5 is not in [1, 2, 3, 4].` this is okay, because this is a scenario where the user needs to make a correction to the file ✔️
> 
> > out of range year, file cannot be submitted because: `File contains data from Jan 1 - Mar 31, which belongs to Fiscal Year 5202, Quarter 2. Adjust your search parameters or upload a different file. Need help?` ✔️
> 
> > out of range year + out of range qtr, file can be submitted but rejected because: `HEADER Item 5 (quarter): 9 is not in [1, 2, 3, 4].` this is okay, because this is a scenario where the user needs to make a correction to the file ✔️
> 
> * Does the function handle cases where the file content is empty or malformed?
> 
> > if empty file, can't submit file because `Request failed with status code 400: The submitted file is empty.` ✔️
> 
> > if no header, then file can be submitted but is rejected because `Your file does not begin with a HEADER record.` ✔️
> 
> * How does it handle multiple headers (with different qtrs)
> 
> > file will be handled according to the first record in the file ✔️
> 
> * How does it handle different file encodings? [reference](https://github.com/raft-tech/TANF-app/pull/3438#pullrequestreview-2601855052))
> 
> > see [here](https://github.com/raft-tech/TANF-app/pull/3501#discussion_r2012846191) for edge case that previously worked ⚠️

@ADPennington Awesome! I think we can address the encoding issue in a follow-on ticket (I'll make it and link it here shortly). 

We missed covering release notes in the past couple syncs, but I can go ahead and get started with a draft! There were a few other issues (#3161, #3478, any material we want to start staging for FRA or ahead of regional staff Project Updates meeting) I wanted to confirm if we want to do a release notes for, but we can approve or nix them in the review cycle.

**Comment by raftmsohani** [comment]
@ADPennington , @lhuxraft Created #3612 as a follow up ticket

**Review by ADPennington — APPROVED** [review]
thanks @raftmsohani  -- looks great! 🚀 

- test notes [here](https://github.com/raft-tech/TANF-app/pull/3501#pullrequestreview-2715194626)
